resource/model.go

```diff
@@ -0,0 +1,140 @@
+package resource
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+
+	// "github.com/jhump/protoreflect/desc".
+	"github.com/pkg/errors"
+)
+
+type (
+	// ModelFamilyName is the model family.
+	ModelFamilyName string
+
+	// ModelName is the name of a specific model within a family.
+	ModelName string
+)
+
+// ModelFamilyDefaultName is the name "default".
+const ModelFamilyDefaultName = ModelFamilyName("builtin")
+
+var (
+	// ModelFamilyDefault is the rdk:builtin model family for built-in resources.
+	ModelFamilyDefault       = ModelFamily{ResourceNamespaceRDK, ModelFamilyDefaultName}
+	modelRegexValidator      = regexp.MustCompile(`^([\w-]+):([\w-]+):([\w-]+)$`)
+	shortModelRegexValidator = regexp.MustCompile(`^([\w-]+)$`)
+)
+
+// ModelFamily is a family of related models.
+type ModelFamily struct {
+	Namespace   Namespace       `json:"namespace"`
+	ModelFamily ModelFamilyName `json:"model_family"`
+}
+
+// NewModelFamily creates a new ModelFamily based on parameters passed in.
+func NewModelFamily(namespace Namespace, family ModelFamilyName) ModelFamily {
+	return ModelFamily{namespace, family}
+}
+
+// Validate ensures that important fields exist and are valid.
+func (f ModelFamily) Validate() error {
+	if f.Namespace == "" {
+		return errors.New("model namespace field for resource missing")
```

Comment:
I followed what resource.Type and resource.Subtype do for validate, so that it remained consistent with that paradigm. Also, as there are no "paths" available here, that utility doesn't quite work. Wouldn't be opposed to handling this better in both places though... but as I went to look into that, I found the resource.*.Validate() methods are only ever entered one place, from the robotClient. They're never called outside of tests otherwise, at least as far as I can tell. Not sure what to do about that.